//---------------------------------------------------------------------
// Copyright 2020 National Instruments
//
// Licensed under the MIT License;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://mit-license.org/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//---------------------------------------------------------------------

//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option java_multiple_files = true;
option java_package = "labview.queryserver";
option java_outer_classname = "QueryServer";
option objc_class_prefix = "LVQS";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package queryserver;

//---------------------------------------------------------------------
// The QueryServer service definition.
//---------------------------------------------------------------------
service QueryServer {
  rpc Invoke (InvokeRequest) returns (InvokeResponse) {}
  rpc Query (QueryRequest) returns (QueryResponse) {}
  rpc Register (RegistrationRequest) returns (stream ServerEvent) {}
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeRequest {
  string command = 1;
  string parameter = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeResponse {
  int32 status = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryRequest {
  string query = 1;
}

//---------------------------------------------------------------------
// Result message from the Query RPC method
//---------------------------------------------------------------------
message QueryResponse {
  // The string result of the server query
  string message = 1;
  // status result of the server query
  int32 status = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message RegistrationRequest {
  string eventName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ServerEvent {
  string eventData = 1;
  int32 serverId = 2;
  int32 status = 3;
}
